neqo-transport/src/send_stream.rs

```diff
@@ -602,20 +571,43 @@ impl SendStream {
     }
 
     pub fn send(&mut self, buf: &[u8]) -> Res<usize> {
-        let sent = match self.state {
-            SendStreamState::Ready => {
-                let mut send_buf = TxBuffer::new();
-                let sent = send_buf.send(buf);
-                self.state.transition(SendStreamState::Send { send_buf });
-                sent
+        if let SendStreamState::Ready = self.state {
+            self.state.transition(SendStreamState::Send {
+                send_buf: TxBuffer::new(),
+            });
+        }
+
+        let stream_credit_avail = self.credit_avail();
+        let conn_credit_avail = self.flow_mgr.borrow().conn_credit_avail();
+        let credit_avail = min(stream_credit_avail, conn_credit_avail);
+        let can_send_bytes = min(credit_avail, buf.len() as u64);
+
+        if !buf.is_empty() && can_send_bytes == 0 {
+            // We had some bytes to send but were blocked by flow
+            // control.
+            assert!(stream_credit_avail == 0 || conn_credit_avail == 0);
+            if stream_credit_avail == 0 {
```

Comment:
We should send the blocked signal any time that `buf.len() > can_send_bytes`, not just when we are completely dry.  Otherwise, we won't be sending the signal often enough.  With this code, writes that are truncated don't trigger the blocked signal, but they really should.

Reply:
Makes sense.